File: server/src/utils/errorCodes.js
// ─────────────────────────────────────────────
// Centralised error codes
// ─────────────────────────────────────────────

export const ERRORS = {
  // ── Authentication ───────────────────────────
  AUTH_001: { code: 'AUTH_001', status: 401, message: 'Authentication token is required' },
  AUTH_002: { code: 'AUTH_002', status: 401, message: 'Token has expired' },
  AUTH_003: { code: 'AUTH_003', status: 401, message: 'Token is invalid' },
  AUTH_004: { code: 'AUTH_004', status: 403, message: 'Insufficient permissions' },
  AUTH_005: { code: 'AUTH_005', status: 401, message: 'System client not found' },
  AUTH_006: { code: 'AUTH_006', status: 403, message: 'System client is inactive' },
  AUTH_007: { code: 'AUTH_007', status: 403, message: 'Endpoint not permitted for this client' },

  // ── Validation ───────────────────────────────
  VAL_001: { code: 'VAL_001', status: 400, message: 'Required field missing' },
  VAL_002: { code: 'VAL_002', status: 400, message: 'Invalid date format (expected YYYY-MM-DD)' },
  VAL_003: { code: 'VAL_003', status: 400, message: 'Invalid enum value' },
  VAL_004: { code: 'VAL_004', status: 400, message: 'Value out of allowed range' },
  VAL_005: { code: 'VAL_005', status: 409, message: 'Duplicate record' },
  VAL_006: { code: 'VAL_006', status: 400, message: 'Referenced record not found' },
  VAL_007: { code: 'VAL_007', status: 400, message: 'CSV missing required columns' },
  VAL_008: { code: 'VAL_008', status: 413, message: 'File too large (max 20 MB)' },
  VAL_009: { code: 'VAL_009', status: 400, message: 'Invalid file type (CSV only)' },
  VAL_010: { code: 'VAL_010', status: 400, message: 'Persistency month must be 13, 25, 37, 49, or 61' },

  // ── Business Rules ───────────────────────────
  BUS_001: { code: 'BUS_001', status: 422, message: 'Program is not in ACTIVE status' },
  BUS_002: { code: 'BUS_002', status: 409, message: 'Overlapping program date range for this channel' },
  BUS_003: { code: 'BUS_003', status: 422, message: 'Cannot modify APPROVED or PAID incentive result' },
  BUS_004: { code: 'BUS_004', status: 404, message: 'Agent not found in hierarchy system' },
  BUS_005: { code: 'BUS_005', status: 409, message: 'Incentive already calculated for this period' },
  BUS_006: { code: 'BUS_006', status: 422, message: 'No payout rules defined for program' },
  BUS_007: { code: 'BUS_007', status: 422, message: 'No KPI rules defined for program' },
  BUS_008: { code: 'BUS_008', status: 422, message: 'Agent license has expired' },
  BUS_009: { code: 'BUS_009', status: 422, message: 'Product is not active' },
  BUS_010: { code: 'BUS_010', status: 422, message: 'Cannot approve — persistency gate failed' },

  // ── Integration ──────────────────────────────
  INT_001: { code: 'INT_001', status: 502, message: 'SFTP connection failed' },
  INT_002: { code: 'INT_002', status: 404, message: 'SFTP file not found' },
  INT_003: { code: 'INT_003', status: 409, message: 'File already processed (duplicate filename)' },
  INT_004: { code: 'INT_004', status: 422, message: 'Staging validation failed' },
  INT_005: { code: 'INT_005', status: 502, message: 'Hierarchy API unreachable' },
  INT_006: { code: 'INT_006', status: 504, message: 'Penta API request timed out' },
  INT_007: { code: 'INT_007', status: 500, message: 'Export file generation failed' },
  INT_008: { code: 'INT_008', status: 422, message: 'SAP FICO format error' },
  INT_009: { code: 'INT_009', status: 422, message: 'Oracle AP format error' },

  // ── Calculation ──────────────────────────────
  CALC_001: { code: 'CALC_001', status: 422, message: 'No performance data found for agent and period' },
  CALC_002: { code: 'CALC_002', status: 422, message: 'No incentive rate defined for product and channel' },
  CALC_003: { code: 'CALC_003', status: 409, message: 'Calculation already in progress for this program' },
  CALC_004: { code: 'CALC_004', status: 422, message: 'Target value is zero (division error)' },
  CALC_005: { code: 'CALC_005', status: 500, message: 'Agent hierarchy path is corrupted' },
};

// ─────────────────────────────────────────────
// Helper — build a standard error response body
// ─────────────────────────────────────────────

/**
 * Build a consistent API error response object.
 *
 * @param {string} code  One of the keys in ERRORS (e.g. 'AUTH_001').
 * @param {*}      [details=null]  Optional extra context (field name, list of
 *                                  missing columns, upstream error message, etc.).
 * @returns {{ success: false, error: string, code: string, details: * }}
 */
export const apiError = (code, details = null) => ({
  success: false,
  error: ERRORS[code].message,
  code,
  details,
});

File: api/Utils/ErrorCodes.cs
namespace IncentiveApi.Utils;

public record ErrorEntry(string Code, int Status, string Message);

public static class ErrorCodes
{
    // ── Authentication ───────────────────────────
    public static readonly ErrorEntry AUTH_001 = new("AUTH_001", 401, "Authentication token is required");
    public static readonly ErrorEntry AUTH_002 = new("AUTH_002", 401, "Token has expired");
    public static readonly ErrorEntry AUTH_003 = new("AUTH_003", 401, "Token is invalid");
    public static readonly ErrorEntry AUTH_004 = new("AUTH_004", 403, "Insufficient permissions");
    public static readonly ErrorEntry AUTH_005 = new("AUTH_005", 401, "System client not found");
    public static readonly ErrorEntry AUTH_006 = new("AUTH_006", 403, "System client is inactive");
    public static readonly ErrorEntry AUTH_007 = new("AUTH_007", 403, "Endpoint not permitted for this client");

    // ── Validation ───────────────────────────────
    public static readonly ErrorEntry VAL_001 = new("VAL_001", 400, "Required field missing");
    public static readonly ErrorEntry VAL_002 = new("VAL_002", 400, "Invalid date format (expected YYYY-MM-DD)");
    public static readonly ErrorEntry VAL_003 = new("VAL_003", 400, "Invalid enum value");
    public static readonly ErrorEntry VAL_004 = new("VAL_004", 400, "Value out of allowed range");
    public static readonly ErrorEntry VAL_005 = new("VAL_005", 409, "Duplicate record");
    public static readonly ErrorEntry VAL_006 = new("VAL_006", 400, "Referenced record not found");
    public static readonly ErrorEntry VAL_007 = new("VAL_007", 400, "CSV missing required columns");
    public static readonly ErrorEntry VAL_008 = new("VAL_008", 413, "File too large (max 20 MB)");
    public static readonly ErrorEntry VAL_009 = new("VAL_009", 400, "Invalid file type (CSV only)");
    public static readonly ErrorEntry VAL_010 = new("VAL_010", 400, "Persistency month must be 13, 25, 37, 49, or 61");

    // ── Business Rules ───────────────────────────
    public static readonly ErrorEntry BUS_001 = new("BUS_001", 422, "Program is not in ACTIVE status");
    public static readonly ErrorEntry BUS_002 = new("BUS_002", 409, "Overlapping program date range for this channel");
    public static readonly ErrorEntry BUS_003 = new("BUS_003", 422, "Cannot modify APPROVED or PAID incentive result");
    public static readonly ErrorEntry BUS_004 = new("BUS_004", 404, "Agent not found in hierarchy system");
    public static readonly ErrorEntry BUS_005 = new("BUS_005", 409, "Incentive already calculated for this period");
    public static readonly ErrorEntry BUS_006 = new("BUS_006", 422, "No payout rules defined for program");
    public static readonly ErrorEntry BUS_007 = new("BUS_007", 422, "No KPI rules defined for program");
    public static readonly ErrorEntry BUS_008 = new("BUS_008", 422, "Agent license has expired");
    public static readonly ErrorEntry BUS_009 = new("BUS_009", 422, "Product is not active");
    public static readonly ErrorEntry BUS_010 = new("BUS_010", 422, "Cannot approve — persistency gate failed");

    // ── Integration ──────────────────────────────
    public static readonly ErrorEntry INT_001 = new("INT_001", 502, "SFTP connection failed");
    public static readonly ErrorEntry INT_002 = new("INT_002", 404, "SFTP file not found");
    public static readonly ErrorEntry INT_003 = new("INT_003", 409, "File already processed (duplicate filename)");
    public static readonly ErrorEntry INT_004 = new("INT_004", 422, "Staging validation failed");
    public static readonly ErrorEntry INT_005 = new("INT_005", 502, "Hierarchy API unreachable");
    public static readonly ErrorEntry INT_006 = new("INT_006", 504, "Penta API request timed out");
    public static readonly ErrorEntry INT_007 = new("INT_007", 500, "Export file generation failed");
    public static readonly ErrorEntry INT_008 = new("INT_008", 422, "SAP FICO format error");
    public static readonly ErrorEntry INT_009 = new("INT_009", 422, "Oracle AP format error");

    // ── Calculation ──────────────────────────────
    public static readonly ErrorEntry CALC_001 = new("CALC_001", 422, "No performance data found for agent and period");
    public static readonly ErrorEntry CALC_002 = new("CALC_002", 422, "No incentive rate defined for product and channel");
    public static readonly ErrorEntry CALC_003 = new("CALC_003", 409, "Calculation already in progress for this program");
    public static readonly ErrorEntry CALC_004 = new("CALC_004", 422, "Target value is zero (division error)");
    public static readonly ErrorEntry CALC_005 = new("CALC_005", 500, "Agent hierarchy path is corrupted");
}

public class ApiException : Exception
{
    public ErrorEntry ErrorEntry { get; }
    public object? Details { get; }

    public ApiException(ErrorEntry errorEntry, object? details = null)
        : base(errorEntry.Message)
    {
        ErrorEntry = errorEntry;
        Details = details;
    }

    public static ApiException From(ErrorEntry entry, object? details = null) => new(entry, details);
}
